File: model_manifest.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Model manifest schema validation and definitions for on-device models."""

import copy
import math
import re
from typing import Any, Dict, List

ALLOWED_SCHEMA_VERSIONS = {1}
ALLOWED_FAMILIES = {'gemma'}
ALLOWED_ADAPTERS = {'litert-lm'}
ALLOWED_FORMATS = {'litertlm'}
ALLOWED_LANGUAGES = {'en', 'ja', 'zh', 'fr', 'de', 'sv'}

MODEL_ID_REGEX = re.compile(r'^[a-z0-9\-]+$')
VERSION_REGEX = re.compile(r'^[a-z0-9.\-]+$')
SHA256_REGEX = re.compile(r'^[a-f0-9]{64}$')
GCS_GENERATION_REGEX = re.compile(r'^[0-9]+$')

PUBLIC_TOP_LEVEL_KEYS = {
    'schemaVersion',
    'modelId',
    'version',
    'displayName',
    'family',
    'adapterId',
    'format',
    'sizeBytes',
    'sha256',
    'gcsGeneration',
    'capabilities',
    'requirements',
    'generation',
}

NESTED_KEYS = {
    'capabilities': {
        'textGeneration', 'languages', 'maxInputTokens', 'maxOutputTokens'
    },
    'requirements': {
        'webgpu', 'minimumDeviceMemoryGB', 'minimumFreeStorageBytes'
    },
    'generation': {'temperature', 'topP', 'maxOutputTokens'},
}


class ManifestValidationError(ValueError):
  """Raised when a model manifest fails schema or security validation."""
  pass


def validate_manifest(raw: Dict[str, Any],
                      allow_private_fields: bool = False) -> Dict[str, Any]:
  """Validates a model manifest dictionary.

  Args:
    raw: Raw manifest dictionary to validate.
    allow_private_fields: If True, allows private server-only fields (gcsBucket, gcsObject).

  Returns:
    A normalized copy of the validated manifest dictionary.

  Raises:
    ManifestValidationError: If the manifest fails schema or safety rules.
  """
  if not isinstance(raw, dict):
    raise ManifestValidationError('Manifest must be a dictionary')

  allowed_keys = set(PUBLIC_TOP_LEVEL_KEYS)
  if allow_private_fields:
    allowed_keys.update({'gcsBucket', 'gcsObject'})

  unknown_keys = set(raw.keys()) - allowed_keys
  if unknown_keys:
    raise ManifestValidationError(
        f'Unknown keys in manifest: {sorted(unknown_keys)}')

  for section, section_keys in NESTED_KEYS.items():
    value = raw.get(section)
    if isinstance(value, dict):
      unknown_nested = set(value) - section_keys
      if unknown_nested:
        raise ManifestValidationError(
            f'Unknown keys in {section}: {sorted(unknown_nested)}')

  # Check for prohibited executable or URL fields
  for key, value in raw.items():
    if isinstance(value, str) and ('http://' in value or 'https://' in value or
                                   'javascript:' in value):
      raise ManifestValidationError(
          f'Field "{key}" must not contain URL schemes')

  # 1. schemaVersion
  schema_version = raw.get('schemaVersion')
  if type(schema_version
         ) is not int or schema_version not in ALLOWED_SCHEMA_VERSIONS:
    raise ManifestValidationError(
        f'Invalid or unsupported schemaVersion: {schema_version}')

  # 2. modelId
  model_id = raw.get('modelId')
  if not isinstance(model_id, str) or not MODEL_ID_REGEX.match(model_id):
    raise ManifestValidationError(f'Invalid modelId: {model_id}')

  # 3. version
  version = raw.get('version')
  if not isinstance(version, str) or not VERSION_REGEX.match(version):
    raise ManifestValidationError(f'Invalid version: {version}')

  # 4. displayName
  display_name = raw.get('displayName')
  if not isinstance(display_name, str) or not display_name.strip():
    raise ManifestValidationError('displayName must be a non-empty string')

  # 5. family
  family = raw.get('family')
  if not isinstance(family, str) or family not in ALLOWED_FAMILIES:
    raise ManifestValidationError(f'Unsupported model family: {family}')

  # 6. adapterId
  adapter_id = raw.get('adapterId')
  if not isinstance(adapter_id, str) or adapter_id not in ALLOWED_ADAPTERS:
    raise ManifestValidationError(
        f'Unsupported or dangerous adapterId: {adapter_id}')

  # 7. format
  model_format = raw.get('format')
  if not isinstance(model_format, str) or model_format not in ALLOWED_FORMATS:
    raise ManifestValidationError(f'Unsupported model format: {model_format}')

  # 8. sizeBytes
  size_bytes = raw.get('sizeBytes')
  if type(
      size_bytes) is not int or size_bytes <= 0 or size_bytes > 20_000_000_000:
    raise ManifestValidationError(
        f'sizeBytes must be between 1 and 20,000,000,000: {size_bytes}')

  # 9. sha256
  sha256 = raw.get('sha256')
  if not isinstance(sha256, str):
    raise ManifestValidationError('sha256 must be a string')
  sha256_lower = sha256.lower()
  if not SHA256_REGEX.match(sha256_lower):
    raise ManifestValidationError(
        f'sha256 must be a 64-character hex digest: {sha256}')

  # 10. gcsGeneration
  gcs_generation = raw.get('gcsGeneration')
  if not isinstance(gcs_generation,
                    str) or not GCS_GENERATION_REGEX.match(gcs_generation):
    raise ManifestValidationError(
        f'gcsGeneration must be a non-empty numeric string: {gcs_generation}')

  # 11. capabilities
  capabilities = raw.get('capabilities')
  if not isinstance(capabilities, dict):
    raise ManifestValidationError('capabilities must be a dictionary')
  if not capabilities.get('textGeneration') is True:
    raise ManifestValidationError('capabilities.textGeneration must be True')
  languages = capabilities.get('languages')
  if not isinstance(languages, list) or not languages:
    raise ManifestValidationError(
        'capabilities.languages must be a non-empty list')
  for lang in languages:
    if not isinstance(lang, str) or lang not in ALLOWED_LANGUAGES:
      raise ManifestValidationError(
          f'capabilities.languages contains unsupported language: {lang}')
  max_input_tokens = capabilities.get('maxInputTokens')
  if type(max_input_tokens) is not int or max_input_tokens <= 0:
    raise ManifestValidationError(
        'capabilities.maxInputTokens must be a positive integer')
  max_output_tokens = capabilities.get('maxOutputTokens')
  if type(max_output_tokens) is not int or max_output_tokens <= 0:
    raise ManifestValidationError(
        'capabilities.maxOutputTokens must be a positive integer')

  # 12. requirements
  requirements = raw.get('requirements')
  if not isinstance(requirements, dict):
    raise ManifestValidationError('requirements must be a dictionary')
  if not requirements.get('webgpu') is True:
    raise ManifestValidationError('requirements.webgpu must be True')
  min_ram = requirements.get('minimumDeviceMemoryGB')
  if (isinstance(min_ram, bool) or not isinstance(min_ram, (int, float)) or
      not math.isfinite(min_ram) or min_ram <= 0):
    raise ManifestValidationError(
        'requirements.minimumDeviceMemoryGB must be positive')
  min_storage = requirements.get('minimumFreeStorageBytes')
  if type(min_storage) is not int or min_storage < size_bytes:
    raise ManifestValidationError(
        f'requirements.minimumFreeStorageBytes ({min_storage}) must be >= sizeBytes ({size_bytes})'
    )

  # 13. generation
  generation = raw.get('generation')
  if not isinstance(generation, dict):
    raise ManifestValidationError('generation must be a dictionary')
  temp = generation.get('temperature')
  if (isinstance(temp, bool) or not isinstance(temp, (int, float)) or
      not math.isfinite(temp) or temp < 0.0 or temp > 2.0):
    raise ManifestValidationError(
        'generation.temperature must be between 0.0 and 2.0')
  top_p = generation.get('topP')
  if (isinstance(top_p, bool) or not isinstance(top_p, (int, float)) or
      not math.isfinite(top_p) or top_p < 0.0 or top_p > 1.0):
    raise ManifestValidationError('generation.topP must be between 0.0 and 1.0')
  gen_max_output = generation.get('maxOutputTokens')
  if type(gen_max_output) is not int or gen_max_output <= 0:
    raise ManifestValidationError(
        'generation.maxOutputTokens must be a positive integer')

  # Private fields validation if allowed
  if allow_private_fields:
    gcs_bucket = raw.get('gcsBucket')
    if gcs_bucket is not None and (not isinstance(gcs_bucket, str) or
                                   not gcs_bucket.strip()):
      raise ManifestValidationError('gcsBucket must be a non-empty string')
    gcs_object = raw.get('gcsObject')
    if gcs_object is not None and (not isinstance(gcs_object, str) or
                                   not gcs_object.strip()):
      raise ManifestValidationError('gcsObject must be a non-empty string')

  validated = copy.deepcopy(raw)
  validated['sha256'] = sha256_lower
  return validated


def get_public_manifest(manifest: Dict[str, Any]) -> Dict[str, Any]:
  """Strips private backend-only fields from a manifest before returning to clients."""
  return {
      k: copy.deepcopy(v)
      for k, v in manifest.items()
      if k in PUBLIC_TOP_LEVEL_KEYS
  }

File: src/on-device/model-manifest.ts
/**
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export interface ModelCapabilities {
  textGeneration: true;
  languages: string[];
  maxInputTokens: number;
  maxOutputTokens: number;
}

export interface ModelRequirements {
  webgpu: true;
  minimumDeviceMemoryGB: number;
  minimumFreeStorageBytes: number;
}

export interface ModelGenerationConfig {
  temperature: number;
  topP: number;
  maxOutputTokens: number;
}

export interface ModelManifest {
  schemaVersion: 1;
  modelId: string;
  version: string;
  displayName: string;
  family: 'gemma';
  adapterId: 'litert-lm';
  format: 'litertlm';
  sizeBytes: number;
  sha256: string;
  gcsGeneration: string;
  capabilities: ModelCapabilities;
  requirements: ModelRequirements;
  generation: ModelGenerationConfig;
}

export class ManifestValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ManifestValidationError';
  }
}

const ALLOWED_FAMILIES = new Set(['gemma']);
const ALLOWED_ADAPTERS = new Set(['litert-lm']);
const ALLOWED_FORMATS = new Set(['litertlm']);
const ALLOWED_LANGUAGES = new Set(['en', 'ja', 'zh', 'fr', 'de', 'sv']);

const MODEL_ID_REGEX = /^[a-z0-9-]+$/;
const VERSION_REGEX = /^[a-z0-9.-]+$/;
const SHA256_REGEX = /^[a-f0-9]{64}$/;
const GCS_GENERATION_REGEX = /^[0-9]+$/;

function containsControlCharacter(value: string): boolean {
  return [...value].some(character => {
    const codePoint = character.codePointAt(0) ?? 0;
    return codePoint <= 31 || codePoint === 127;
  });
}

const ALLOWED_TOP_LEVEL_KEYS = new Set([
  'schemaVersion',
  'modelId',
  'version',
  'displayName',
  'family',
  'adapterId',
  'format',
  'sizeBytes',
  'sha256',
  'gcsGeneration',
  'capabilities',
  'requirements',
  'generation',
]);

const ALLOWED_CAPABILITY_KEYS = new Set([
  'textGeneration',
  'languages',
  'maxInputTokens',
  'maxOutputTokens',
]);
const ALLOWED_REQUIREMENT_KEYS = new Set([
  'webgpu',
  'minimumDeviceMemoryGB',
  'minimumFreeStorageBytes',
]);
const ALLOWED_GENERATION_KEYS = new Set([
  'temperature',
  'topP',
  'maxOutputTokens',
]);

function rejectUnknownKeys(
  value: Record<string, unknown>,
  allowed: Set<string>,
  section: string,
): void {
  for (const key of Object.keys(value)) {
    if (!allowed.has(key)) {
      throw new ManifestValidationError(`Unknown key in ${section}: ${key}`);
    }
  }
}

/**
 * Validates raw data as a safe, canonical ModelManifest.
 * Rejects unknown fields, malicious URL schemes, and invalid numeric bounds.
 */
export function validateModelManifest(raw: unknown): ModelManifest {
  if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) {
    throw new ManifestValidationError('Manifest must be a non-null object');
  }

  const record = raw as Record<string, unknown>;

  // Check unknown top-level keys
  for (const key of Object.keys(record)) {
    if (!ALLOWED_TOP_LEVEL_KEYS.has(key)) {
      throw new ManifestValidationError(`Unknown key in manifest: ${key}`);
    }
  }

  // Reject URL schemes or script content in any string field
  for (const [key, value] of Object.entries(record)) {
    if (typeof value === 'string') {
      if (
        value.includes('http://') ||
        value.includes('https://') ||
        value.includes('javascript:')
      ) {
        throw new ManifestValidationError(
          `Field "${key}" must not contain URL schemes`,
        );
      }
    }
  }

  // 1. schemaVersion
  if (record.schemaVersion !== 1) {
    throw new ManifestValidationError(
      `Invalid schemaVersion: ${record.schemaVersion}`,
    );
  }

  // 2. modelId
  if (
    typeof record.modelId !== 'string' ||
    record.modelId.length > 64 ||
    !MODEL_ID_REGEX.test(record.modelId)
  ) {
    throw new ManifestValidationError(`Invalid modelId: ${record.modelId}`);
  }

  // 3. version
  if (
    typeof record.version !== 'string' ||
    record.version.length > 64 ||
    !VERSION_REGEX.test(record.version)
  ) {
    throw new ManifestValidationError(`Invalid version: ${record.version}`);
  }

  // 4. displayName
  if (
    typeof record.displayName !== 'string' ||
    record.displayName.trim() === '' ||
    record.displayName.length > 128 ||
    containsControlCharacter(record.displayName)
  ) {
    throw new ManifestValidationError(
      'displayName must be a non-empty safe string of at most 128 characters',
    );
  }

  // 5. family
  if (
    typeof record.family !== 'string' ||
    !ALLOWED_FAMILIES.has(record.family)
  ) {
    throw new ManifestValidationError(
      `Unsupported model family: ${record.family}`,
    );
  }

  // 6. adapterId
  if (
    typeof record.adapterId !== 'string' ||
    !ALLOWED_ADAPTERS.has(record.adapterId)
  ) {
    throw new ManifestValidationError(
      `Unsupported or dangerous adapterId: ${record.adapterId}`,
    );
  }

  // 7. format
  if (
    typeof record.format !== 'string' ||
    !ALLOWED_FORMATS.has(record.format)
  ) {
    throw new ManifestValidationError(
      `Unsupported model format: ${record.format}`,
    );
  }

  // 8. sizeBytes
  if (
    typeof record.sizeBytes !== 'number' ||
    !Number.isInteger(record.sizeBytes) ||
    record.sizeBytes <= 0 ||
    record.sizeBytes > 20_000_000_000
  ) {
    throw new ManifestValidationError(
      `sizeBytes must be an integer between 1 and 20,000,000,000: ${record.sizeBytes}`,
    );
  }

  // 9. sha256
  if (typeof record.sha256 !== 'string') {
    throw new ManifestValidationError('sha256 must be a string');
  }
  const sha256Lower = record.sha256.toLowerCase();
  if (!SHA256_REGEX.test(sha256Lower)) {
    throw new ManifestValidationError(
      `sha256 must be a 64-character hex digest: ${record.sha256}`,
    );
  }

  // 10. gcsGeneration
  if (
    typeof record.gcsGeneration !== 'string' ||
    !GCS_GENERATION_REGEX.test(record.gcsGeneration) ||
    /^0+$/.test(record.gcsGeneration)
  ) {
    throw new ManifestValidationError(
      `gcsGeneration must be a non-empty numeric string: ${record.gcsGeneration}`,
    );
  }

  // 11. capabilities
  const cap = record.capabilities as Record<string, unknown> | undefined;
  if (typeof cap !== 'object' || cap === null || Array.isArray(cap)) {
    throw new ManifestValidationError('capabilities must be an object');
  }
  rejectUnknownKeys(cap, ALLOWED_CAPABILITY_KEYS, 'capabilities');
  if (cap.textGeneration !== true) {
    throw new ManifestValidationError(
      'capabilities.textGeneration must be true',
    );
  }
  if (!Array.isArray(cap.languages) || cap.languages.length === 0) {
    throw new ManifestValidationError(
      'capabilities.languages must be a non-empty array',
    );
  }
  for (const lang of cap.languages) {
    if (typeof lang !== 'string' || !ALLOWED_LANGUAGES.has(lang)) {
      throw new ManifestValidationError(
        `capabilities.languages contains unsupported language: ${lang}`,
      );
    }
  }
  if (new Set(cap.languages).size !== cap.languages.length) {
    throw new ManifestValidationError(
      'capabilities.languages must not contain duplicates',
    );
  }
  if (
    typeof cap.maxInputTokens !== 'number' ||
    !Number.isInteger(cap.maxInputTokens) ||
    cap.maxInputTokens <= 0 ||
    cap.maxInputTokens > 32768
  ) {
    throw new ManifestValidationError(
      'capabilities.maxInputTokens must be between 1 and 32768',
    );
  }
  if (
    typeof cap.maxOutputTokens !== 'number' ||
    !Number.isInteger(cap.maxOutputTokens) ||
    cap.maxOutputTokens <= 0 ||
    cap.maxOutputTokens > 4096
  ) {
    throw new ManifestValidationError(
      'capabilities.maxOutputTokens must be between 1 and 4096',
    );
  }

  // 12. requirements
  const req = record.requirements as Record<string, unknown> | undefined;
  if (typeof req !== 'object' || req === null || Array.isArray(req)) {
    throw new ManifestValidationError('requirements must be an object');
  }
  rejectUnknownKeys(req, ALLOWED_REQUIREMENT_KEYS, 'requirements');
  if (req.webgpu !== true) {
    throw new ManifestValidationError('requirements.webgpu must be true');
  }
  if (
    typeof req.minimumDeviceMemoryGB !== 'number' ||
    !Number.isFinite(req.minimumDeviceMemoryGB) ||
    req.minimumDeviceMemoryGB <= 0 ||
    req.minimumDeviceMemoryGB > 1024
  ) {
    throw new ManifestValidationError(
      'requirements.minimumDeviceMemoryGB must be between 0 and 1024',
    );
  }
  if (
    typeof req.minimumFreeStorageBytes !== 'number' ||
    !Number.isInteger(req.minimumFreeStorageBytes) ||
    req.minimumFreeStorageBytes < record.sizeBytes ||
    req.minimumFreeStorageBytes > 100_000_000_000
  ) {
    throw new ManifestValidationError(
      `requirements.minimumFreeStorageBytes (${req.minimumFreeStorageBytes}) must be >= sizeBytes (${record.sizeBytes})`,
    );
  }

  // 13. generation
  const gen = record.generation as Record<string, unknown> | undefined;
  if (typeof gen !== 'object' || gen === null || Array.isArray(gen)) {
    throw new ManifestValidationError('generation must be an object');
  }
  rejectUnknownKeys(gen, ALLOWED_GENERATION_KEYS, 'generation');
  if (
    typeof gen.temperature !== 'number' ||
    !Number.isFinite(gen.temperature) ||
    gen.temperature < 0 ||
    gen.temperature > 2.0
  ) {
    throw new ManifestValidationError(
      'generation.temperature must be between 0.0 and 2.0',
    );
  }
  if (
    typeof gen.topP !== 'number' ||
    !Number.isFinite(gen.topP) ||
    gen.topP < 0 ||
    gen.topP > 1.0
  ) {
    throw new ManifestValidationError(
      'generation.topP must be between 0.0 and 1.0',
    );
  }
  if (
    typeof gen.maxOutputTokens !== 'number' ||
    !Number.isInteger(gen.maxOutputTokens) ||
    gen.maxOutputTokens <= 0 ||
    gen.maxOutputTokens > cap.maxOutputTokens
  ) {
    throw new ManifestValidationError(
      'generation.maxOutputTokens must be positive and no greater than capabilities.maxOutputTokens',
    );
  }

  return {
    schemaVersion: 1,
    modelId: record.modelId,
    version: record.version,
    displayName: record.displayName,
    family: record.family as 'gemma',
    adapterId: record.adapterId as 'litert-lm',
    format: record.format as 'litertlm',
    sizeBytes: record.sizeBytes,
    sha256: sha256Lower,
    gcsGeneration: record.gcsGeneration,
    capabilities: {
      textGeneration: true,
      languages: [...cap.languages],
      maxInputTokens: cap.maxInputTokens,
      maxOutputTokens: cap.maxOutputTokens,
    },
    requirements: {
      webgpu: true,
      minimumDeviceMemoryGB: req.minimumDeviceMemoryGB,
      minimumFreeStorageBytes: req.minimumFreeStorageBytes,
    },
    generation: {
      temperature: gen.temperature,
      topP: gen.topP,
      maxOutputTokens: gen.maxOutputTokens,
    },
  };
}
